File: content/develop/use-cases/job-queue/rust/src/worker.rs
//! Mock background worker for the job-queue demo.
//!
//! A worker pulls jobs off the queue, simulates work by sleeping for a
//! configurable latency, and either completes the job, fails it, or
//! intentionally hangs to simulate a worker crash that the reclaimer must
//! recover from. This is the demo-side stand-in for whatever real work your
//! application would run in the background (sending emails, transcoding
//! video, calling third-party webhooks, etc.).

use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Arc;
use std::time::Duration;

use rand::Rng;
use serde_json::json;
use tokio::sync::{watch, Mutex};
use tokio::task::JoinHandle;

use crate::job_queue::{ClaimedJob, RedisJobQueue};

/// Knobs shared by every worker in a pool. Held under a mutex so the demo
/// server can adjust them at runtime.
#[derive(Clone, Debug)]
pub struct WorkerConfig {
    pub work_latency_ms: u64,
    pub fail_rate: f64,
    pub hang_rate: f64,
}

impl Default for WorkerConfig {
    fn default() -> Self {
        Self {
            work_latency_ms: 400,
            fail_rate: 0.0,
            hang_rate: 0.0,
        }
    }
}

/// A single background worker task.
struct Worker {
    name: String,
    queue: RedisJobQueue,
    config: Arc<Mutex<WorkerConfig>>,
    stop_rx: watch::Receiver<bool>,
    processed: Arc<AtomicI64>,
}

impl Worker {
    fn new(
        name: String,
        queue: RedisJobQueue,
        config: Arc<Mutex<WorkerConfig>>,
        stop_rx: watch::Receiver<bool>,
    ) -> Self {
        Self {
            name,
            queue,
            config,
            stop_rx,
            processed: Arc::new(AtomicI64::new(0)),
        }
    }

    async fn run(self) {
        while !*self.stop_rx.borrow() {
            let job = match self.queue.claim(500).await {
                Ok(Some(job)) => job,
                Ok(None) => continue,
                Err(err) => {
                    eprintln!("[{}] claim error: {}", self.name, err);
                    tokio::time::sleep(Duration::from_millis(200)).await;
                    continue;
                }
            };

            self.process(job).await;
        }
    }

    async fn process(&self, job: ClaimedJob) {
        let cfg = self.config.lock().await.clone();
        let outcome = pick_outcome(cfg.fail_rate, cfg.hang_rate);
        tokio::time::sleep(Duration::from_millis(cfg.work_latency_ms)).await;

        match outcome {
            Outcome::Hang => {
                // Simulate a worker that crashed mid-job: don't complete,
                // don't fail. The reclaimer will move this job back to
                // pending once the visibility timeout elapses.
            }
            Outcome::Fail => {
                let msg = format!("{} simulated failure", self.name);
                if let Err(err) = self.queue.fail(&job, &msg).await {
                    eprintln!("[{}] fail error: {}", self.name, err);
                }
            }
            Outcome::Ok => {
                self.processed.fetch_add(1, Ordering::Relaxed);
                let result = json!({
                    "worker": self.name,
                    "echo": job.payload,
                    "attempts": job.attempts,
                });
                if let Err(err) = self.queue.complete(&job, result).await {
                    eprintln!("[{}] complete error: {}", self.name, err);
                }
            }
        }
    }
}

enum Outcome {
    Hang,
    Fail,
    Ok,
}

fn pick_outcome(fail_rate: f64, hang_rate: f64) -> Outcome {
    let roll: f64 = rand::thread_rng().gen();
    if roll < hang_rate {
        Outcome::Hang
    } else if roll < hang_rate + fail_rate {
        Outcome::Fail
    } else {
        Outcome::Ok
    }
}

/// A handle to a running worker task plus its shared `processed` counter.
struct WorkerHandle {
    handle: JoinHandle<()>,
    processed: Arc<AtomicI64>,
}

/// A pool of named worker tasks that can be started and stopped.
pub struct WorkerPool {
    queue: RedisJobQueue,
    config: Arc<Mutex<WorkerConfig>>,
    size: Mutex<usize>,
    state: Mutex<PoolState>,
}

struct PoolState {
    handles: Vec<WorkerHandle>,
    stop_tx: Option<watch::Sender<bool>>,
    historical_processed: i64,
}

impl WorkerPool {
    pub fn new(queue: RedisJobQueue, size: usize, config: WorkerConfig) -> Self {
        Self {
            queue,
            config: Arc::new(Mutex::new(config)),
            size: Mutex::new(size),
            state: Mutex::new(PoolState {
                handles: Vec::new(),
                stop_tx: None,
                historical_processed: 0,
            }),
        }
    }

    /// Replace the in-memory size. Workers are not started/stopped here;
    /// call `start` afterwards.
    pub async fn resize(&self, size: usize) {
        let mut current = self.size.lock().await;
        *current = size;
    }

    /// Stop any running workers and spawn a fresh batch at the configured
    /// size. Idempotent: calling it on an already-running pool restarts it
    /// with the latest config.
    pub async fn start(&self) {
        self.stop().await;

        let size = *self.size.lock().await;
        if size == 0 {
            return;
        }

        let (stop_tx, stop_rx) = watch::channel(false);
        let mut handles: Vec<WorkerHandle> = Vec::with_capacity(size);

        for i in 0..size {
            let name = format!("worker-{}", i + 1);
            let worker = Worker::new(
                name,
                self.queue.clone(),
                self.config.clone(),
                stop_rx.clone(),
            );
            let processed = worker.processed.clone();
            let handle = tokio::spawn(worker.run());
            handles.push(WorkerHandle { handle, processed });
        }

        let mut state = self.state.lock().await;
        state.handles = handles;
        state.stop_tx = Some(stop_tx);
    }

    /// Signal workers to stop and wait for them to finish their current
    /// `claim()`/`process()` cycle. Their processed counts are folded into
    /// `historical_processed` so the running total persists across restarts.
    pub async fn stop(&self) {
        let mut state = self.state.lock().await;
        if let Some(tx) = state.stop_tx.take() {
            let _ = tx.send(true);
        }
        let handles = std::mem::take(&mut state.handles);
        let mut historical = state.historical_processed;
        drop(state);

        for h in handles {
            historical += h.processed.load(Ordering::Relaxed);
            let _ = h.handle.await;
        }

        let mut state = self.state.lock().await;
        state.historical_processed = historical;
    }

    /// Number of currently-running worker tasks.
    pub async fn running(&self) -> usize {
        let state = self.state.lock().await;
        state.handles.iter().filter(|h| !h.handle.is_finished()).count()
    }

    /// Sum of jobs processed by the live pool plus any workers from previous
    /// runs (mirrors the Python `total_processed` accumulation).
    pub async fn total_processed(&self) -> i64 {
        let state = self.state.lock().await;
        let live: i64 = state
            .handles
            .iter()
            .map(|h| h.processed.load(Ordering::Relaxed))
            .sum();
        state.historical_processed + live
    }

    /// Zero every live worker's processed counter and clear the historical
    /// accumulation, so `total_processed` returns 0 immediately after.
    pub async fn reset_processed(&self) {
        let mut state = self.state.lock().await;
        for h in &state.handles {
            h.processed.store(0, Ordering::Relaxed);
        }
        state.historical_processed = 0;
    }

    pub async fn config_snapshot(&self) -> WorkerConfig {
        self.config.lock().await.clone()
    }

    pub async fn configure(
        &self,
        work_latency_ms: Option<u64>,
        fail_rate: Option<f64>,
        hang_rate: Option<f64>,
    ) {
        let mut cfg = self.config.lock().await;
        if let Some(latency) = work_latency_ms {
            cfg.work_latency_ms = latency;
        }
        if let Some(rate) = fail_rate {
            cfg.fail_rate = rate.clamp(0.0, 1.0);
        }
        if let Some(rate) = hang_rate {
            cfg.hang_rate = rate.clamp(0.0, 1.0);
        }
    }
}

File: content/develop/use-cases/job-queue/go/worker.go
// Mock background worker for the job-queue demo.
//
// A worker pulls jobs off the queue, simulates work by sleeping for a
// configurable latency, and either completes the job, fails it, or
// intentionally hangs to simulate a worker crash that the reclaimer must
// recover from. This is the demo-side stand-in for whatever real work
// your application would run in the background (sending emails,
// transcoding video, calling third-party webhooks, etc.).
package jobqueue

import (
	"context"
	"fmt"
	"math/rand"
	"sync"
	"sync/atomic"
	"time"
)

// Worker is a single background goroutine that drains a Redis job queue.
type Worker struct {
	name  string
	queue *RedisJobQueue

	mu            sync.Mutex
	workLatencyMs int
	failRate      float64
	hangRate      float64

	processed atomic.Int64

	// Lifecycle
	runMu  sync.Mutex
	cancel context.CancelFunc
	done   chan struct{}
	rng    *rand.Rand
}

// NewWorker creates a new Worker bound to the given queue.
func NewWorker(name string, queue *RedisJobQueue, workLatencyMs int, failRate, hangRate float64) *Worker {
	return &Worker{
		name:          name,
		queue:         queue,
		workLatencyMs: workLatencyMs,
		failRate:      failRate,
		hangRate:      hangRate,
		rng:           rand.New(rand.NewSource(time.Now().UnixNano())),
	}
}

// Configure updates the worker's runtime knobs. Safe to call while the
// worker is running.
func (w *Worker) Configure(workLatencyMs int, failRate, hangRate float64) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.workLatencyMs = workLatencyMs
	w.failRate = failRate
	w.hangRate = hangRate
}

func (w *Worker) snapshot() (int, float64, float64) {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.workLatencyMs, w.failRate, w.hangRate
}

// Start runs the worker loop in a new goroutine. If the worker is already
// running, Start is a no-op. If a previous run is shutting down it waits
// for it to finish first.
func (w *Worker) Start(parent context.Context) {
	w.runMu.Lock()
	defer w.runMu.Unlock()

	if w.done != nil {
		select {
		case <-w.done:
			// Previous run finished; we can start a new one.
		default:
			// Still running.
			return
		}
	}

	ctx, cancel := context.WithCancel(parent)
	w.cancel = cancel
	done := make(chan struct{})
	w.done = done
	go w.run(ctx, done)
}

// Stop signals the worker to shut down. It does not block.
func (w *Worker) Stop() {
	w.runMu.Lock()
	cancel := w.cancel
	w.runMu.Unlock()
	if cancel != nil {
		cancel()
	}
}

// IsAlive reports whether the worker's goroutine is currently running.
func (w *Worker) IsAlive() bool {
	w.runMu.Lock()
	done := w.done
	w.runMu.Unlock()
	if done == nil {
		return false
	}
	select {
	case <-done:
		return false
	default:
		return true
	}
}

// Processed returns the number of jobs this worker has successfully
// completed since it was created.
func (w *Worker) Processed() int64 {
	return w.processed.Load()
}

// ResetProcessed zeroes this worker's processed counter.
func (w *Worker) ResetProcessed() {
	w.processed.Store(0)
}

func (w *Worker) run(ctx context.Context, done chan struct{}) {
	defer close(done)
	for {
		select {
		case <-ctx.Done():
			return
		default:
		}
		job, err := w.queue.Claim(ctx, 500)
		if err != nil {
			// On error, back off briefly and retry unless cancelled.
			select {
			case <-ctx.Done():
				return
			case <-time.After(100 * time.Millisecond):
				continue
			}
		}
		if job == nil {
			continue
		}
		w.process(ctx, job)
	}
}

func (w *Worker) process(ctx context.Context, job *ClaimedJob) {
	// Decide outcome up front so the latency reflects "work was tried".
	outcome := w.pickOutcome()
	latency, _, _ := w.snapshot()
	if latency > 0 {
		select {
		case <-ctx.Done():
			return
		case <-time.After(time.Duration(latency) * time.Millisecond):
		}
	}

	switch outcome {
	case "hang":
		// Simulate a worker that crashed mid-job: don't complete, don't
		// fail. The reclaimer will move this job back to pending once the
		// visibility timeout elapses.
		return
	case "fail":
		_, _ = w.queue.Fail(ctx, job, fmt.Sprintf("%s simulated failure", w.name))
		return
	default:
		// "ok"
		ok, err := w.queue.Complete(ctx, job, map[string]any{
			"worker":   w.name,
			"echo":     job.Payload,
			"attempts": job.Attempts,
		})
		if err == nil && ok {
			w.processed.Add(1)
		}
	}
}

func (w *Worker) pickOutcome() string {
	w.mu.Lock()
	roll := w.rng.Float64()
	failRate := w.failRate
	hangRate := w.hangRate
	w.mu.Unlock()
	if roll < hangRate {
		return "hang"
	}
	if roll < hangRate+failRate {
		return "fail"
	}
	return "ok"
}

// WorkerPool is a pool of named Worker goroutines that can be started and
// stopped together.
type WorkerPool struct {
	queue *RedisJobQueue

	mu            sync.Mutex
	workers       []*Worker
	workLatencyMs int
	failRate      float64
	hangRate      float64

	parentCtx context.Context
}

// NewWorkerPool builds a pool sized to size workers. Pass a context that
// outlives the pool (cancelling it will stop every worker).
func NewWorkerPool(ctx context.Context, queue *RedisJobQueue, size, workLatencyMs int, failRate, hangRate float64) *WorkerPool {
	p := &WorkerPool{
		queue:         queue,
		workLatencyMs: workLatencyMs,
		failRate:      failRate,
		hangRate:      hangRate,
		parentCtx:     ctx,
	}
	p.Resize(size)
	return p
}

// Resize grows or shrinks the pool to the requested size. Removed workers
// are stopped.
func (p *WorkerPool) Resize(size int) {
	p.mu.Lock()
	defer p.mu.Unlock()
	for len(p.workers) < size {
		w := NewWorker(
			fmt.Sprintf("worker-%d", len(p.workers)+1),
			p.queue,
			p.workLatencyMs,
			p.failRate,
			p.hangRate,
		)
		p.workers = append(p.workers, w)
	}
	for len(p.workers) > size {
		idx := len(p.workers) - 1
		p.workers[idx].Stop()
		p.workers = p.workers[:idx]
	}
}

// Start applies the current configuration to every worker and starts any
// that aren't already running.
func (p *WorkerPool) Start() {
	p.mu.Lock()
	defer p.mu.Unlock()
	for _, w := range p.workers {
		w.Configure(p.workLatencyMs, p.failRate, p.hangRate)
		w.Start(p.parentCtx)
	}
}

// Stop signals every worker in the pool to shut down.
func (p *WorkerPool) Stop() {
	p.mu.Lock()
	defer p.mu.Unlock()
	for _, w := range p.workers {
		w.Stop()
	}
}

// Running counts the workers whose goroutine is currently active.
func (p *WorkerPool) Running() int {
	p.mu.Lock()
	defer p.mu.Unlock()
	n := 0
	for _, w := range p.workers {
		if w.IsAlive() {
			n++
		}
	}
	return n
}

// TotalProcessed sums Processed() across the pool.
func (p *WorkerPool) TotalProcessed() int64 {
	p.mu.Lock()
	defer p.mu.Unlock()
	var n int64
	for _, w := range p.workers {
		n += w.Processed()
	}
	return n
}

// ResetProcessed zeroes every worker's processed counter.
func (p *WorkerPool) ResetProcessed() {
	p.mu.Lock()
	defer p.mu.Unlock()
	for _, w := range p.workers {
		w.ResetProcessed()
	}
}

// WorkLatencyMs returns the currently configured latency.
func (p *WorkerPool) WorkLatencyMs() int {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.workLatencyMs
}

// FailRate returns the currently configured failure rate.
func (p *WorkerPool) FailRate() float64 {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.failRate
}

// HangRate returns the currently configured hang rate.
func (p *WorkerPool) HangRate() float64 {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.hangRate
}

// Configure updates one or more pool-wide knobs. Pass nil to leave a
// value unchanged.
func (p *WorkerPool) Configure(workLatencyMs *int, failRate, hangRate *float64) {
	p.mu.Lock()
	defer p.mu.Unlock()
	if workLatencyMs != nil {
		v := *workLatencyMs
		if v < 0 {
			v = 0
		}
		p.workLatencyMs = v
	}
	if failRate != nil {
		v := *failRate
		if v < 0 {
			v = 0
		}
		if v > 1 {
			v = 1
		}
		p.failRate = v
	}
	if hangRate != nil {
		v := *hangRate
		if v < 0 {
			v = 0
		}
		if v > 1 {
			v = 1
		}
		p.hangRate = v
	}
	for _, w := range p.workers {
		w.Configure(p.workLatencyMs, p.failRate, p.hangRate)
	}
}
